### PR TITLE
* #14: Improved user input validation.

### LINKED ISSUE #14: Improve user input validation
## Problem

We use the same pattern to validate user input. We should validate database identifiers, credentials and file names in different ways.

### PR DESCRIPTION
**Pull request by AnastasiiaSergienko**
Closes #14

### CODE REVIEW

File: src/main/java/com/exasol/adapter/installer/UserInputValidator.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    private UserInputValidator() {
+    }
+
+    /**
+     * Validate an Exasol identifier.
+     *
+     * @param value the value
+     * @return validated string
+     */
+    public static String validateExasolIdentifier(final String value) {
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/installer/UserInputValidator.java:22]:
```suggestion
    public static String validateExasolIdentifier(final String exasolIdentifier) {
```

File: src/main/java/com/exasol/adapter/installer/UserInputValidator.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        return validateString(value, allowedSpecialCharacters);
+    }
+
+    /**
+     * Validate a name of a file.
+     *
+     * @param value the value
+     * @return validated string
+     */
+    public static String validateFileName(final String value) {
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/installer/UserInputValidator.java:33]:
```suggestion
    public static String validateFileName(final String filename) {
```

File: src/main/java/com/exasol/adapter/installer/UserInputValidator.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        return validateString(value, allowedSpecialCharacters);
+    }
+
+    /**
+     * Validate a path.
+     *
+     * @param value the value
+     * @return validated string
+     */
+    public static String validatePath(final String value) {
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/installer/UserInputValidator.java:44]:
```suggestion
    public static String validatePath(final String path) {
```

Comment (morazow) [reply src/main/java/com/exasol/adapter/installer/UserInputValidator.java:44]: Renamed to keep consistent with other methods. Please check them and do not forget to change the Javadoc-s also.

File: src/main/java/com/exasol/adapter/installer/dialect/DialectPropertyParser.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        } else {
+            return createDialectProperty(property, property.substring(0, firstEqualityMark),
+                    property.substring(firstEqualityMark + 1));
+        }
+    }
+
+    private static InstallerException createException(final String property) {
+        return new InstallerException(ExaError.messageBuilder("E-VS-INSTL-10")
+                .message("Can't parse dialect property: {{property}}", property)
+                .mitigation("Dialect property must have the following format: <PROPERTY>:'<VALUE>'").toString());
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/installer/dialect/DialectPropertyParser.java:32]: Should ':' be '=' in the mitigation? I could not see the relation between first equality mark and the error mitigation.

File: src/main/java/com/exasol/adapter/installer/dialect/DialectProperty.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    @Override
+    public boolean equals(final Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/installer/dialect/DialectProperty.java:50]:
```suggestion
        if (!(o isinstanceof DialectProperty)) {
            return false;
        }
```

Comment (morazow) [reply src/main/java/com/exasol/adapter/installer/dialect/DialectProperty.java:50]: getClass version is usually autogenerated, Effective Java Item 8 suggests to use isinstanceof.

File: src/main/java/com/exasol/adapter/installer/Installer.java [hunk 6]
```diff
@@ -232,32 +222,32 @@ public InstallerBuilder jdbcDriver(final JdbcDriver jdbcDriver) {
...
         public InstallerBuilder sourcePassword(final String sourcePassword) {
-            this.sourcePassword = InputString.validate(sourcePassword);
+            this.sourcePassword = UserInputValidator.validateLiteralString(sourcePassword);
             return this;
         }
 
         public InstallerBuilder exaIp(final String exaIp) {
-            this.exaIp = InputString.validate(exaIp);
+            this.exaIp = exaIp;
             return this;
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/installer/Installer.java:251]: Do we need validation for this? Previously it was also validated. Same for username, password and bucket password. Maybe username still need validation.

File: src/main/java/com/exasol/adapter/installer/Installer.java [hunk 7]
```diff
@@ -135,55 +136,45 @@ private String prepareVirtualSchemaStatement() {
...
         }
         return createVirtualSchemaStatement.toString();
     }
 
     private String prepareAdapterScriptStatement(final List<String> jarNames) {
         final StringBuilder adapterScript = new StringBuilder( //
-                "CREATE OR REPLACE JAVA ADAPTER SCRIPT " + this.exaSchemaName + "." + this.exaAdapterName + " AS"
-                        + LINE_SEPARATOR //
+                "CREATE OR REPLACE JAVA ADAPTER SCRIPT \"" + this.exaSchemaName + "\".\"" + this.exaAdapterName
+                        + "\" AS" + LINE_SEPARATOR //
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/installer/Installer.java:150]: :+1:

**Review by morazow — CHANGES_REQUESTED** [review]

File: src/main/java/com/exasol/adapter/installer/Installer.java (hunk 6 above)
Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/adapter/installer/Installer.java:251]: I added validation to the values that are used in the SQL statementss. ExaIP, username and password are only used to create a connection to Exasol via Java code (with JDBC driver). So I think we don't really need to validate them. Bucket password is also used only in code

File: src/main/java/com/exasol/adapter/installer/UserInputValidator.java (hunk 3 above)
Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/adapter/installer/UserInputValidator.java:44]: I think in the method validatePropertyKey it's not nice to have a parameter with "value" name... Maybe it's okay to keep them different?

File: src/main/java/com/exasol/adapter/installer/dialect/DialectProperty.java (hunk 5 above)
Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/adapter/installer/dialect/DialectProperty.java:50]: The methods are generated by my IDE

File: src/main/java/com/exasol/adapter/installer/dialect/DialectPropertyParser.java (hunk 4 above)
Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/adapter/installer/dialect/DialectPropertyParser.java:32]: You're right, it's a typo

**Review by morazow — APPROVED** [review]

**Pull request merged**